### PR TITLE
feat(http_body_limit): add initial support for DefaultBodyLimit

### PR DESCRIPTION
**Pull request by etolbakov**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Use `DefaultBodyLimit` layer to avoid the default 2MB limit on the body

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link
https://github.com/GreptimeTeam/greptimedb/issues/1842

### CODE REVIEW

**Comment by etolbakov** [comment]
@evenyag @zyy17 @waynexia @killme2008 
Could I kindly request review for this draft? Does that match your expectations?
I decided to pick this up as I feel this feature could be quite handy. 

Once you agree with the approach, I'm happy to move forward and add more tests (at the moment there are only in [frontend.rs](https://github.com/GreptimeTeam/greptimedb/pull/1860/files#diff-f020b56aa77f4b8cc1d72a4f9f7215c6cb9c0a31c73a3e6082dfefb569046296R264)) and add the parsing logic similar to the one to [datanode.rs](https://github.com/GreptimeTeam/greptimedb/pull/1860/files#diff-012667a4e46e13aada8d971ded24c8078764cfcea1bca454179613b27ca1402dR169)

I also encourage @ItIsOHM to watch this ticket and share suggestions as well

**Comment by ItIsOHM** [comment]
Ah thanks @etolbakov for looping me in! This is very helpful to a beginner like me :D

**Comment by killme2008** [comment]
> @evenyag @zyy17 @waynexia @killme2008 Could I kindly request review for this draft? Does that match your expectations? I decided to pick this up as I feel this feature could be quite handy.
> 
> Once you agree with the approach, I'm happy to move forward and add more tests (at the moment there are only in [frontend.rs](https://github.com/GreptimeTeam/greptimedb/pull/1860/files#diff-f020b56aa77f4b8cc1d72a4f9f7215c6cb9c0a31c73a3e6082dfefb569046296R264)) and add the parsing logic similar to the one to [datanode.rs](https://github.com/GreptimeTeam/greptimedb/pull/1860/files#diff-012667a4e46e13aada8d971ded24c8078764cfcea1bca454179613b27ca1402dR169)
> 
> I also encourage @ItIsOHM to watch this ticket and share suggestions as well

It's great, please go ahead!
One of our community users has requested this feature, their payload size may be greater than 6GB.

File: src/servers/src/http.rs [hunk 1]
```diff
@@ -141,6 +145,7 @@ impl Default for HttpOptions {
             addr: "127.0.0.1:4000".to_string(),
             timeout: Duration::from_secs(30),
             disable_dashboard: false,
+            http_body_maximum_size: ReadableSize::from_str("6G").unwrap(),
```
Comment (evenyag) [inline src/servers/src/http.rs:148]: suggestion: `6G` is so large. What about `64M`?

**Review by evenyag — COMMENTED** [review]
Could you also please update the examples in `standalone.example.toml` and `frontend.example.toml`.

https://github.com/GreptimeTeam/greptimedb/blob/b466ef6cb620730453822a6fa2be57437d97e3bd/config/standalone.example.toml#L7-L11

File: src/servers/src/http.rs [hunk 2]
```diff
@@ -133,6 +135,8 @@ pub struct HttpOptions {
 
     #[serde(skip)]
     pub disable_dashboard: bool,
+
+    pub http_body_maximum_size: ReadableSize,
```
[the remark this answers is not in the document]
Comment (etolbakov) [reply src/servers/src/http.rs:139]: Agree, thanks for the suggestion!

**Comment by etolbakov** [comment]
@evenyag 
Could you please take a look when you have time?

File: config/standalone.example.toml [hunk 3]
```diff
@@ -9,6 +9,9 @@ enable_memory_catalog = false
 addr = "127.0.0.1:4000"
 # HTTP request timeout, 30s by default.
 timeout = "30s"
+# HTTP request body limit, 64Mb by default.
+# the following units are supported: B, KB, KiB, MB, MiB, GB, GiB, TB, TiB, PB, PiB
+body_limit = "64Mb"
```
Comment (etolbakov) [inline config/standalone.example.toml:14]: just realized that `Mb` is not in the supported units 😅
will make a follow up adjustment after the feedback on the PR

File: config/frontend.example.toml [hunk 4]
```diff
@@ -9,6 +9,7 @@ retry_interval_millis = 5000
 [http_options]
 addr = "127.0.0.1:4000"
 timeout = "30s"
+body_limit = "64Mb"
```
Comment (etolbakov) [inline config/frontend.example.toml:12]: same as below

File: src/servers/src/http.rs [hunk 5]
```diff
@@ -546,7 +551,9 @@ impl HttpServer {
                     .layer(HandleErrorLayer::new(handle_error))
                     .layer(TraceLayer::new_for_http())
                     .layer(TimeoutLayer::new(self.options.timeout))
-                    .layer(DefaultBodyLimit::max(DEFAULT_BODY_LIMIT))
+                    .layer(DefaultBodyLimit::max(
+                        self.options.body_limit.0.try_into().unwrap(),
```
Comment (waynexia) [inline src/servers/src/http.rs:555]: not sure if `unwrap()` is suitable here... @evenyag

**Review by waynexia — APPROVED** [review]
Great work! Other places LGTM 👍

File: src/servers/src/http.rs (hunk 5 above)
Comment (etolbakov) [reply src/servers/src/http.rs:555]: yeah... I was not sure if I need to use `snafu` error during conversion. Happy to adjust if needed

File: src/servers/src/http.rs [hunk 6]
```diff
@@ -105,7 +107,7 @@ pub(crate) async fn query_context_from_db(
 pub const HTTP_API_VERSION: &str = "v1";
 pub const HTTP_API_PREFIX: &str = "/v1/";
 /// Default http body limit (64M).
-const DEFAULT_BODY_LIMIT: usize = 64 * 1024 * 1024;
+const DEFAULT_BODY_LIMIT: &str = "64M";
```
Comment (evenyag) [inline src/servers/src/http.rs:110]: suggestion: `ReadableSize::mb()` is a constant function, so you could try
```suggestion
const DEFAULT_BODY_LIMIT: ReadableSize = ReadableSize::mb(64);
```

Refer to
https://github.com/GreptimeTeam/greptimedb/blob/884731a2c807fe5d0e5ac86bb5863d4a06bfeafd/src/common/base/src/readable_size.rs#L45

File: src/servers/src/http.rs (hunk 5 above)
Comment (evenyag) [reply src/servers/src/http.rs:555]: What about using the default value as a fallback? Casting DEFAULT_BODY_LIMIT to usize should be safe.
```suggestion
                        self.options.body_limit.0.try_into().unwrap_or_else(|| DEFAULT_BODY_LIMIT.as_bytes() as usize),
```

File: src/servers/src/http.rs [hunk 7]
```diff
@@ -143,6 +147,7 @@ impl Default for HttpOptions {
             addr: "127.0.0.1:4000".to_string(),
             timeout: Duration::from_secs(30),
             disable_dashboard: false,
+            body_limit: ReadableSize::from_str(DEFAULT_BODY_LIMIT).unwrap(),
```
Comment (evenyag) [inline src/servers/src/http.rs:150]:
```suggestion
            body_limit: DEFAULT_BODY_LIMIT,
```

**Review by evenyag — COMMENTED** [review]
You could use `ReadableSize` as a constant. It might make error handling easier.

**Comment by etolbakov** [comment]
@evenyag I've updated the PR according to the suggestions.
Could you please take a look once again?

**Review by evenyag — APPROVED** [review]
LGTM ❤️

**Pull request merged**